File: app/src/main/java/com/gytxtx/openjbd/protocol/JbdFrame.java
package com.gytxtx.openjbd.protocol;

import java.util.Arrays;

public final class JbdFrame {
    public final int command;
    public final int status;
    public final byte[] payload;

    private JbdFrame(int command, int status, byte[] payload) {
        this.command = command;
        this.status = status;
        this.payload = payload;
    }

    public static JbdFrame parse(byte[] frame) throws JbdParseException {
        if (frame.length < 7) {
            throw new JbdParseException("Frame too short");
        }
        if ((frame[0] & 0xFF) != 0xDD || (frame[frame.length - 1] & 0xFF) != 0x77) {
            throw new JbdParseException("Bad frame boundary");
        }
        int command = frame[1] & 0xFF;
        int status = frame[2] & 0xFF;
        int length = frame[3] & 0xFF;
        if (frame.length != length + 7) {
            throw new JbdParseException("Bad frame length");
        }
        byte[] payload = Arrays.copyOfRange(frame, 4, 4 + length);
        int expected = JbdCommands.checksum((byte) status, length, payload);
        int actual = ((frame[frame.length - 3] & 0xFF) << 8) | (frame[frame.length - 2] & 0xFF);
        if (expected != actual) {
            throw new JbdParseException("Bad checksum");
        }
        return new JbdFrame(command, status, payload);
    }
}

File: app/src/main/kotlin/com/gytxtx/openjbd/protocol/JbdFrame.kt
package com.gytxtx.openjbd.protocol

class JbdFrame
private constructor(
    @JvmField val command: Int,
    @JvmField val status: Int,
    @JvmField val payload: ByteArray
) {
    companion object {
        @JvmStatic
        fun parse(frame: ByteArray): JbdFrame {
            if (frame.size < 7) {
                throw JbdParseException("Frame too short")
            }
            if ((frame[0].toInt() and 0xFF) != 0xDD || (frame[frame.size - 1].toInt() and 0xFF) != 0x77) {
                throw JbdParseException("Bad frame boundary")
            }
            val command = frame[1].toInt() and 0xFF
            val status = frame[2].toInt() and 0xFF
            val length = frame[3].toInt() and 0xFF
            if (frame.size != length + 7) {
                throw JbdParseException("Bad frame length")
            }
            val payload = frame.copyOfRange(4, 4 + length)
            val expected = JbdCommands.checksum(status.toByte(), length, payload)
            val actual =
                ((frame[frame.size - 3].toInt() and 0xFF) shl 8) or (frame[frame.size - 2].toInt() and 0xFF)
            if (expected != actual) {
                throw JbdParseException("Bad checksum")
            }
            return JbdFrame(command, status, payload)
        }
    }
}
